### PR TITLE
Implement `FloatingTypes`

### PR DESCRIPTION
**Pull request by lcartey**
## Description

This PR implements the `FloatingTypes` package.

I made a number of assumptions when implementing this package:

 1. Floating-point types are implemented on the target platform using IEEE 754.
 2. The size of the mantissa for 4 byte and 8 byte floats are 23 bits and 52 bits, respectively.
 3. The floating-point implementation supports +/- infinity.

As I believe our supported compiler/platform combinations adhere to these requirements.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - `FLP32-C`
     - `FLP34-C`
     - `FLP36-C`
     - `FLP37-C`
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [ ] Have all the relevant rule package description files been checked in?
 - [ ] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [ ] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [ ] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [ ] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [ ] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/cpp/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

RULE                 : A0-4-4
QUERY                : UncheckedRangeDomainPoleErrors
SUITE                : AUTOSAR
TEST_PASS            : True
TEST_DIFFERENCE      : 
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
PACKAGE              : TypeRanges


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

QUERY                : UncheckedFloatingPointConversion
TEST_DIFFERENCE      : 
COMPILE_PASS         : True
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
PACKAGE              : FloatingTypes
RULE                 : FLP34-C

QUERY                : IntToFloatPreservePrecision
TEST_DIFFERENCE      : 
COMPILE_PASS         : True
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
PACKAGE              : FloatingTypes
RULE                 : FLP36-C

QUERY                : MemcmpUsedToCompareFloats
TEST_DIFFERENCE      : 
COMPILE_PASS         : True
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
PACKAGE              : FloatingTypes
RULE                 : FLP37-C

QUERY                : UncheckedRangeDomainPoleErrors
TEST_DIFFERENCE      : 
COMPILE_PASS         : True
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
PACKAGE              : FloatingTypes
RULE                 : FLP32-C


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/c/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

RULE                 : FLP34-C
SUITE                : CERT-C
QUERY                : UncheckedFloatingPointConversion
PACKAGE              : FloatingTypes
TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
COMPILE_ERROR_OUTPUT : 

RULE                 : FLP36-C
SUITE                : CERT-C
QUERY                : IntToFloatPreservePrecision
PACKAGE              : FloatingTypes
TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
COMPILE_ERROR_OUTPUT : 

RULE                 : FLP37-C
SUITE                : CERT-C
QUERY                : MemcmpUsedToCompareFloats
PACKAGE              : FloatingTypes
TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
COMPILE_ERROR_OUTPUT : 

RULE                 : FLP32-C
SUITE                : CERT-C
QUERY                : UncheckedRangeDomainPoleErrors
PACKAGE              : FloatingTypes
TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
COMPILE_ERROR_OUTPUT : 


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/cpp/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

COMPILE_ERROR_OUTPUT : 
QUERY                : UncheckedRangeDomainPoleErrors
TEST_DIFFERENCE      : 
SUITE                : AUTOSAR
PACKAGE              : TypeRanges
COMPILE_PASS         : True
RULE                 : A0-4-4
TEST_PASS            : True


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! Matrix Testing for this PR has been **completed**. If no reports were posted it means this PR does not contain things that need matrix testing!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/cpp/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

PACKAGE              : TypeRanges
QUERY                : UncheckedRangeDomainPoleErrors
RULE                 : A0-4-4
TEST_PASS            : True
SUITE                : AUTOSAR
COMPILE_ERROR_OUTPUT : 
TEST_DIFFERENCE      : 
COMPILE_PASS         : True


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/cpp/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

TEST_PASS            : True
COMPILE_PASS         : True
COMPILE_ERROR_OUTPUT : 
RULE                 : A0-4-4
PACKAGE              : TypeRanges
TEST_DIFFERENCE      : 
QUERY                : UncheckedRangeDomainPoleErrors
SUITE                : AUTOSAR


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

RULE                 : FLP34-C
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : UncheckedFloatingPointConversion
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
COMPILE_PASS         : True

RULE                 : FLP36-C
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : IntToFloatPreservePrecision
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
COMPILE_PASS         : True

RULE                 : FLP37-C
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : MemcmpUsedToCompareFloats
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
COMPILE_PASS         : True

RULE                 : FLP32-C
TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : UncheckedRangeDomainPoleErrors
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
COMPILE_PASS         : True


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! Matrix Testing for this PR has been **completed**. If no reports were posted it means this PR does not contain things that need matrix testing!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/X86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
QUERY                : UncheckedFloatingPointConversion
PACKAGE              : FloatingTypes
TEST_DIFFERENCE      : 
SUITE                : CERT-C
RULE                 : FLP34-C

TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
QUERY                : IntToFloatPreservePrecision
PACKAGE              : FloatingTypes
TEST_DIFFERENCE      : 
SUITE                : CERT-C
RULE                 : FLP36-C

TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
QUERY                : MemcmpUsedToCompareFloats
PACKAGE              : FloatingTypes
TEST_DIFFERENCE      : 
SUITE                : CERT-C
RULE                 : FLP37-C

TEST_PASS            : True
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
QUERY                : UncheckedRangeDomainPoleErrors
PACKAGE              : FloatingTypes
TEST_DIFFERENCE      : 
SUITE                : CERT-C
RULE                 : FLP32-C


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/cpp/X86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

COMPILE_PASS         : True
TEST_DIFFERENCE      : 
COMPILE_ERROR_OUTPUT : 
PACKAGE              : TypeRanges
QUERY                : UncheckedRangeDomainPoleErrors
SUITE                : AUTOSAR
TEST_PASS            : True
RULE                 : A0-4-4


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **qcc/cpp/AARCH64LE** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

TEST_PASS            : False
TEST_DIFFERENCE      : --- expected
                       +++ actual
                       @@ -1,3 +1,9 @@
                       +| file:///opt/qcc/qnx-sdp/target/qnx7/usr/include/c++/v1/math.h:775:91:775:97 | call to acosf | Domain error in call to acosf: the argument has a range -Infinity...Infinity which is outside the domain of this function (-1.0...1.0). |
                       +| file:///opt/qcc/qnx-sdp/target/qnx7/usr/include/c++/v1/math.h:776:91:776:97 | call to acosl | Domain error in call to acosl: the argument has a range -Infinity...Infinity which is outside the domain of this function (-1.0...1.0). |
                       +| file:///opt/qcc/qnx-sdp/target/qnx7/usr/include/c++/v1/math.h:787:91:787:97 | call to asinf | Domain error in call to asinf: the argument has a range -Infinity...Infinity which is outside the domain of this function (-1.0...1.0). |
                       +| file:///opt/qcc/qnx-sdp/target/qnx7/usr/include/c++/v1/math.h:788:91:788:97 | call to asinl | Domain error in call to asinl: the argument has a range -Infinity...Infinity which is outside the domain of this function (-1.0...1.0). |
                       +| file:///opt/qcc/qnx-sdp/target/qnx7/usr/include/c++/v1/math.h:1086:92:1086:99 | call to atanhf | Domain error in call to atanhf: the argument has a range -Infinity...Infinity which is outside the domain of this function (-1.0...1.0). |
                       +| file:///opt/qcc/qnx-sdp/target/qnx7/usr/include/c++/v1/math.h:1087:92:1087:99 | call to atanhl | Domain error in call to atanhl: the argument has a range -Infinity...Infinity which is outside the domain of this function (-1.0...1.0). |
                        | test.cpp:4:3:4:6 | call to acos | Domain error in call to acos: the argument has a range -1.1...-1.1 which is outside the domain of this function (-1.0...1.0). |
                        | test.cpp:8:3:8:6 | call to acos | Domain error in call to acos: the argument has a range 1.1...1.1 which is outside the domain of this function (-1.0...1.0). |
                        | test.cpp:9:3:9:6 | call to asin | Domain error in call to asin: the argument has a range -1.1...-1.1 which is outside the domain of this function (-1.0...1.0). |
                       [1/1 comp 30.5s eval 1.2s] FAILED(RESULT) /__w/codeql-coding-standards-release-engineering/codeql-coding-standards-release-engineering/codeql-coding-standards/cpp/common/test/rules/uncheckedrangedomainpoleerrors/UncheckedRangeDomainPoleErrors.ql
                       
RULE                 : A0-4-4
COMPILE_ERROR_OUTPUT : 
SUITE                : AUTOSAR
COMPILE_PASS         : True
PACKAGE              : TypeRanges
QUERY                : UncheckedRangeDomainPoleErrors


```

**Review by kraiouchkine — COMMENTED** [review]
LGTM other than one note for `FLP34-C`.

File: c/cert/src/rules/FLP34-C/UncheckedFloatingPointConversion.ql [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  not isExcluded(c, FloatingTypesPackage::uncheckedFloatingPointConversionQuery()) and
+  underlyingTypeAfter = c.getUnderlyingType() and
+  not (
+    // Either the upper or lower bound of the expression is outside the range of the new type
+    withinIntegralRange(underlyingTypeAfter, [upperBound(c.getExpr()), lowerBound(c.getExpr())])
+    or
+    // Heuristic - is there are guard the abs value of the float can fit in the precision of an int?
+    exists(GuardCondition gc, FunctionCall log2f, FunctionCall fabsf, Expr precision |
+      // gc.controls(c, false) and
+      log2f.getTarget().hasGlobalOrStdName("log2f") and
```
[the remark this answers is not in the document]
Comment (lcartey) [reply c/cert/src/rules/FLP34-C/UncheckedFloatingPointConversion.ql:80]: Yes, will adjust.

**Comment by lcartey** [comment]
The new compiler test failure for **qcc/cpp/AARCH64LE** is an artefact of moving the A0-4-4 query to a shared implementation. We have an internal ticket tracking updating our test infrastructure to handle this case (as shared queries do not currently exclude results outside the source root).

**Comment by jsinglet** [comment]
🤖 Beep Boop! **qcc/cpp/AARCH64LE** Matrix Testing for this PR won't happen because it is outside of license window!

File: c/cert/src/rules/FLP34-C/UncheckedFloatingPointConversion.ql (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcartey) [reply c/cert/src/rules/FLP34-C/UncheckedFloatingPointConversion.ql]: Fixed.

**Comment by jsinglet** [comment]
🤖 Beep Boop! **qcc/c/AARCH64LE** Matrix Testing for this PR won't happen because it is outside of license window!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/X86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
RULE                 : FLP34-C
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : UncheckedFloatingPointConversion

TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
RULE                 : FLP36-C
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : IntToFloatPreservePrecision

TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
RULE                 : FLP37-C
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : MemcmpUsedToCompareFloats

TEST_PASS            : True
COMPILE_PASS         : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
RULE                 : FLP32-C
COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
QUERY                : UncheckedRangeDomainPoleErrors


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/c/X86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
COMPILE_PASS         : True
RULE                 : FLP34-C
TEST_PASS            : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
QUERY                : UncheckedFloatingPointConversion

COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
COMPILE_PASS         : True
RULE                 : FLP36-C
TEST_PASS            : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
QUERY                : IntToFloatPreservePrecision

COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
COMPILE_PASS         : True
RULE                 : FLP37-C
TEST_PASS            : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
QUERY                : MemcmpUsedToCompareFloats

COMPILE_ERROR_OUTPUT : 
SUITE                : CERT-C
COMPILE_PASS         : True
RULE                 : FLP32-C
TEST_PASS            : True
TEST_DIFFERENCE      : 
PACKAGE              : FloatingTypes
QUERY                : UncheckedRangeDomainPoleErrors


```

**Pull request merged**